pkg/frontend/querymiddleware/querysharding_test.go

```diff
@@ -1635,48 +1640,83 @@ func TestQuerySharding_ShouldReturnErrorInCorrectFormat(t *testing.T) {
 	)
 
 	for _, tc := range []struct {
-		name             string
-		engineDownstream *promql.Engine
-		engineSharding   *promql.Engine
-		expError         error
-		queryable        storage.Queryable
+		name                 string
+		engineType           string
+		engineDownstreamOpts []engineOpt
+		engineShardingOpts   []engineOpt
+		expError             error
+		queryable            storage.Queryable
 	}{
+		// Prometheus engine tests.
+
+		{
+			name:                 "downstream - timeout",
+			engineType:           querier.PrometheusEngine,
+			engineDownstreamOpts: []engineOpt{withTimeout(50 * time.Millisecond)},
+			expError:             apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
+			queryable:            queryableSlow,
+		},
+		{
+			name:                 "downstream - sample limit",
+			engineType:           querier.PrometheusEngine,
+			engineDownstreamOpts: []engineOpt{withMaxSamples(1)},
+			expError:             apierror.New(apierror.TypeExec, "query processing would load too many samples into memory in query execution"),
+		},
+		{
+			name:               "sharding - timeout",
+			engineType:         querier.PrometheusEngine,
+			engineShardingOpts: []engineOpt{withTimeout(50 * time.Millisecond)},
+			expError:           apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
+			queryable:          queryableSlow,
+		},
 		{
-			name:             "downstream - timeout",
-			engineDownstream: engineTimeout,
-			engineSharding:   engine,
-			expError:         apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
-			queryable:        queryableSlow,
+			name:               "downstream - storage internal error",
+			engineType:         querier.PrometheusEngine,
+			engineShardingOpts: []engineOpt{withMaxSamples(1)},
+			queryable:          queryableInternalErr,
+			expError:           apierror.New(apierror.TypeInternal, "some internal error"),
 		},
 		{
-			name:             "sharding - sample limit",
-			engineDownstream: engineSampleLimit,
-			engineSharding:   engine,
-			expError:         apierror.New(apierror.TypeExec, "query processing would load too many samples into memory in query execution"),
+			name:               "downstream - storage prometheus execution error",
+			engineType:         querier.PrometheusEngine,
+			engineShardingOpts: []engineOpt{withMaxSamples(1)},
+			queryable:          queryablePrometheusExecErr,
+			expError:           apierror.Newf(apierror.TypeExec, "expanding series: %s", querier.NewMaxQueryLengthError(744*time.Hour, 720*time.Hour)),
 		},
+
+		// MQE equivalents when applicable. For example, MQE doesn't have a sample limit
+		// and returns different error messages for timeouts when executing a query.
+
 		{
```

Comment:
This test currently fails because of the way that timeouts are checked in the Prometheus engine vs the Mimir engine. In Prometheus, the context is checked for timeouts [after the query has been evaluated](https://github.com/grafana/mimir/blob/cc1c7bfb93e2e4774b58cfc68ff48fedd7265c12/vendor/github.com/prometheus/prometheus/promql/engine.go#L824-L826) and a few other places. In contrast, we only check for timeouts [every 128 series](https://github.com/grafana/mimir/blob/cc1c7bfb93e2e4774b58cfc68ff48fedd7265c12/pkg/streamingpromql/operators/selectors/selector.go#L110-L115). I'm not sure what to do about this. Should we check for timeouts in more places? Remove the test?

Reply:
Another option would be to move our existing check for context cancellation before we increment the series index. That guarantees we check the context at least once for all selectors _and_ after 128 series. WDYT?

Reply:
Swapping the order so we check for context cancellation before we increment the series index sounds good to me.